src/main/webapp/isolate/request/scripts/TicketFactory.js

```diff
@@ -0,0 +1,110 @@
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Creates Tickets for embedding in the request form.
+ * Tickets created by the TicketFactory have their name attributes assigned
+ * programatically, to ensure that multiple tickets can be sent within a single
+ * form.
+ * 
+ * This class __must__ be instantiated in order to function correctly.
+ */
+export class TicketFactory {
+  /**
+   * Construct a TicketFactory.
+   */
+  constructor() {
+    /* The number of tickets created. */
+    this.ticketCount = 0;
+    /* The template for request tickets. */
+    this.ticketTemplate = null;
+  }
+
+  /**
+   * Attempt to cache the Ticket Template, if necessary. If the Ticket Template already
+   * exists, calling this function is a no-op.
+   * 
+   * @return Whether the template has been cached.
+   * @throws Will throw an Error if the server cannot retrieve the
+   *     Ticket Template.
+   */
+  async cacheTicketTemplate() {
+    /* If the ticketTemplate is null or undefined, attempt to fetch it from the
+     * server. */
+    if (this.ticketTemplate === null ||
+        typeof this.ticketTemplate === 'undefined') {
+      const response = await fetch('request-ticket.html');
+
+      /* If the request succeeds, attempt to parse the body of the response */
+      if (response.ok) {
+        const text = await response.text();
+
+        /* If parsing the response body fails, report failure. */
+        if (text === null || typeof text === 'undefined') {
+          return false;
+        }
+        /* If parsing the response body succeeds, cache the ticket template and report success. */
+        else {
+          this.ticketTemplate = text;
+          return true;
+        }
+      }
+      /* The response was not ok. Report failure. */
+      return false;
+    }
+    /* The ticketTemplate has already been cached. Report success. */
+    return true;
```

Comment:
I think this would be more readable if you put this step at the top, and instead of checking that the template is null, check that it isn't null. That way I can clearly see that for a cached ticket the result is returned early.

You can then restructure the rest of the code to say, if the !response.ok return false, otherwise just execute the code for getting and caching the ticket.

Reply:
I've restructured the code now so that else statements are implicit, and so that the short-circuiting behaviour when the ticketTemplate is already cached is more easily understood 👍 Thanks for the guidance in terms of how to improve this piece of code https://github.com/googleinterns/step226-2020/pull/5/commits/441cbbdf3e92f89730e9c21e871e1dd3fb9c4eed https://github.com/googleinterns/step226-2020/pull/5/commits/7ba487129fa20c0a83c9299655ca6a27c29293f0